django/test/selenium.py

```diff
@@ -133,3 +137,86 @@ def disable_implicit_wait(self):
             yield
         finally:
             self.selenium.implicitly_wait(self.implicit_wait)
+
+
+def screenshot_cases(method_names):
+    if isinstance(method_names, str):
+        method_names = method_names.split(",")
+
+    def wrapper(func):
+        func._screenshot_cases = method_names
+        return func
+
+    return wrapper
+
+
+@tag("screenshot")
+class ScreenshotTestCase(SeleniumTestCase):
+    screenshots = False
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        for name, func in list(cls.__dict__.items()):
+            if not cls.screenshots:
+                continue
+            if not isinstance(func, FunctionType):
+                continue
+            if not hasattr(func, "_screenshot_cases"):
+                continue
+
+            delattr(cls, name)
+            _screenshot_cases = getattr(func, "_screenshot_cases")
+            for screenshot_case in _screenshot_cases:
+
+                @wraps(func)
+                def test(self, *args, _func=func, _case=screenshot_case, **kwargs):
+                    getattr(self, _case)()
+                    output = _func(self, *args, **kwargs)
+                    getattr(self, f"disable_{_case}")()
+                    return output
+
+                test.__name__ = f"{name}_{screenshot_case}"
+                test.__qualname__ = f"{test.__qualname__}_{screenshot_case}"
+
+                setattr(cls, test.__name__, test)
+
+    def desktop(self):
+        self.window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(1280, 720)
+
+    def disable_desktop(self):
+        self.selenium.set_window_size(
+            self.window_size["width"], self.window_size["height"]
+        )
```

Comment:
turn these into context managers? Otherwise failures can pollute other tests.

Reply:
Yes! I was doing this going, why is this so ugly?! :joy: thank you :+1: